mathesar/utils/tables.py

```diff
@@ -17,8 +24,43 @@ def get_table_column_types(table):
     return col_types
 
 
+def _gen_default_table_name(schema):
+    table_num = Table.objects.count()
+    name = TABLE_NAME_TEMPLATE % table_num
+    metadata = MetaData(bind=schema._sa_engine, schema=schema.name)
+    metadata.reflect()
+    while '.'.join((schema.name, name)) in metadata.tables:
+        table_num += 1
+        name = TABLE_NAME_TEMPLATE % table_num
+        if len(name) > POSTGRES_NAME_LEN_CAP:
+            name = str(table_num)
+    return name
+
+
+def _gen_base_table_name(schema, base_name):
+    name = base_name
+    metadata = MetaData(bind=schema._sa_engine, schema=schema.name)
+    metadata.reflect()
+    table_num = 1
+    while '.'.join((schema.name, name)) in metadata.tables:
+        name = base_name + f' {table_num}'
+        if len(name) > POSTGRES_NAME_LEN_CAP:
+            base_name = base_name[:-1]
+            name = base_name + f' {table_num}'
+        table_num += 1
+    return name
+
+
 def create_table_from_datafile(data_files, name, schema):
     data_file = data_files[0]
+
+    if not name:
```

Comment:
It might be cleanest to put this in a `pre_save` signal or in the `Table` model's `save` method. That way, you don't have to repeat it in every table creation utility.

Reply:
We need access to the source data file for naming, so unfortunately I don't think we can use the save method. I was able to move the naming logic to the viewset, which I think has the desired affect.